avoids starting tracer for Accumulo 2.1.0

keith-turner opened the pull request:
Trying to start the tracer for Accumulo 2.1.0-SNAPSHOT was causing problems.  Changed code to avoid doing that.

keith-turner commented:
I have only tested this change w/ 2.1.0-SNAPSHOT and it fixed the problem I was seeing.  I have not tried running Muchos w/ 2.0.0 though.

arvindshmicrosoft commented:
For completeness, can you share the details about the original error you encountered?

keith-turner commented:
> For completeness, can you share the details about the original error you encountered?

@arvindshmicrosoft this is the error I saw.

```
TASK [start accumulo 2.0 master, monitor, gc & tracer] **********************************************************************************************************************************************************
Wednesday 05 January 2022  17:53:37 +0000 (0:00:00.766)       0:03:01.386 *****
changed: [ktwal2-0] => (item=master)
changed: [ktwal2-0] => (item=monitor)
changed: [ktwal2-0] => (item=gc)
failed: [ktwal2-0] (item=tracer) => {"ansible_loop_var": "item", "changed": false, "cmd": ["nohup", "/home/centos/install/accumulo-2.1.0-SNAPSHOT/bin/accumulo-service", "tracer", "start"], "delta": "0:00:00.071548", "end": "2022-01-05 17:53:43.238652", "item": "tracer", "msg": "non-zero return code", "rc": 1, "start": "2022-01-05 17:53:43.167104", "stderr": "nohup: ignoring input\nUsage: accumulo-service <service> <command>\n\nServices:\n  gc                     Accumulo garbage collector\n  monitor                Accumulo monitor\n  manager                Accumulo manager\n  master                 Accumulo master (Deprecated)\n  tserver                Accumulo tserver\n  compaction-coordinator Accumulo compaction coordinator (experimental)\n  compactor              Accumulo compactor (experimental)\n\nCommands:\n  start       Starts service\n  stop        Stops service\n  kill        Kills service", "stderr_lines": ["nohup: ignoring input", "Usage: accumulo-service <service> <command>", "", "Services:", "  gc
 Accumulo garbage collector", "  monitor                Accumulo monitor", "  manager                Accumulo manager", "  master                 Accumulo master (Deprecated)", "  tserver                Accumulo tserver", "  compaction-coordinator Accumulo compaction coordinator (experimental)", "  compactor              Accumulo compactor (experimental)", "", "Commands:", "  start       Starts service", "  stop        Stops service", "  kill        Kills service"], "stdout": "Invalid arguments: 'tracer' is an invalid <service>", "stdout_lines": ["Invalid arguments: 'tracer' is an invalid <service>"]}
```

keith-turner commented:
I think the problem was caused by the tracer service being removed in apache/accumulo#2259

arvindshmicrosoft commented on ansible/accumulo.yml:
```diff
@@ -53,16 +53,18 @@
+      when: accumulo_version is version('2.0.0','>=') and accumulo_version is version('2.1.0','<') and not use_systemd
 - hosts: accumulomaster
   tasks:
     - name: "install and start all the accumulo services using systemd"
```
There will be similar set of `when` conditions needed for the systemd unit for tracer. It would be good to address that as well in this PR, or maybe log an Issue to track the future change needed.

brianloss replied:
Yes, that would be needed. If you expand this file, it's on line 73/75 below.

keith-turner replied:
I changed the systemd code later in the file but didn't test it (I did run muchos setup again and that was ok, but did not try enabling systemd).

arvindshmicrosoft replied:
@keith-turner - I did test your latest changes with accumulo-2.1.0-SNAPSHOT built off accumulo main, with systemd enabled. While setup did complete without any errors, the observations from @brianloss need to be addressed sometime - especially the accumulo-cluster-systemd script errors out.

BTW, unrelated to this change, the accumulo-cluster-systemd also errors out due to lack of change to `manager` from the older term. That is an unrelated issue and can be cleaned up separately.

keith-turner replied:
> BTW, unrelated to this change, the accumulo-cluster-systemd also errors out due to lack of change to manager from the older term. That is an unrelated issue and can be cleaned up separately.

I made one small change related to this in daeab54 because I was touching that code.  Looking at the file its pervasive , so would be nicer to do that in its own PR.  I wonder if we can avoid a bunch of if stmts in the template and just have template variable that resolves to manager or master.

ctubbsii commented on ansible/accumulo.yml:
```diff
@@ -53,16 +53,18 @@
 - hosts: accumulomaster
   tasks:
-    - name: "start accumulo 2.0 master, monitor, gc & tracer"
+    - name: "start accumulo 2.0 master, monitor, & gc"
```
```suggestion
    - name: "start accumulo 2.x master, monitor, and gc"
```

ctubbsii commented on ansible/accumulo.yml:
```diff
@@ -53,26 +53,32 @@
       when: accumulo_major_version == '2' and not use_systemd
+    - name: "start accumulo 2.0 tracer"
+      command: "nohup {{ accumulo_home }}/bin/accumulo-service tracer start"
+      when: accumulo_version is version('2.0.0','>=') and accumulo_version is version('2.1.0','<') and not use_systemd
```
To be equivalent to what's in the other task, I think you also need:

```suggestion
      when: accumulo_version is version('2.0.0','>=') and accumulo_version is version('2.1.0','<') and not use_systemd
      register: start_tracer_result
      changed_when: "'Starting' in start_tracer_result.stdout"
```

arvindshmicrosoft replied:
yes, what this would do is to make Ansible report a green "ok" status for that play / task, when tracer is already running.

brianloss reviewed (changes requested):
I looked through all of the code for references to tracer, and it seems there are a few other places where we should clean up references to the tracer for Accumulo >= 2.1...

- ansible/wipe-systemd.yml: the accumulo-tracer service won't exist so it can't be stopped
- ansible/roles/accumulo/files/accumulo-cluster-systemd: this might need to be turned into a jinja template so the tracer sections can be included conditionally
- ansible/roles/accumulo/tasks/main.yml: don't copy the tracers file for >= 2.1.0
- ansible/roles/accumulo/templates/accumulo-env.sh: remove tracer from the accumulo 2.1.0 case

arvindshmicrosoft commented:
> * ansible/wipe-systemd.yml: the accumulo-tracer service won't exist so it can't be stopped

It can be made conditional for sure though, but JFYI it is benign as Ansible will not error when the service does not exist.

> * ansible/roles/accumulo/files/accumulo-cluster-systemd: this might need to be turned into a jinja template so the tracer sections can be included conditionally

This should be ideally addressed in this PR itself. Currently the `accumulo-cluster-systemd` script errors out.

brianloss commented on ansible/roles/accumulo/templates/accumulo-cluster-systemd:
```diff
@@ -75,6 +75,7 @@ function verify_config {
     fi
   fi
 
+{% if accumulo_version is version('2.1.0','<') %}
```
I'm not a jinja expert, and this would totally be a nit, but you can specify a dash after the begin and or before the close. That will remove whitespace. A pattern I've seen before is something like:
```
{%- some jinja conditional %}
stuff
{% end -%}
```
The only difference this would make is how clean the file looks after it's been run through the template engine.

keith-turner replied:
I tried playing around with this feature locally and seeing how it rendered.  I didn't use it, but as a result I moved a few empty lines there were outside the template to inside of it.

brianloss commented on ansible/roles/accumulo/templates/accumulo-cluster-systemd:
```diff
@@ -287,26 +292,35 @@ function stop_all() {
   stop_tservers
 
   echo "Cleaning all server entries in ZooKeeper"
+{% if accumulo_version is version('2.1.0','<') %}
   ${accumulo_cmd} org.apache.accumulo.server.util.ZooZap -master -tservers -tracers
+{% else %}
+  ${accumulo_cmd} org.apache.accumulo.server.util.ZooZap -manager -tservers
+{% endif %}
```
Wondering if this would be better as a single line or not?
```suggestion
  ${accumulo_cmd} org.apache.accumulo.server.util.ZooZap -master -tservers{{ ' -tracers' if accumulo_version is version('2.1.0','<') }}
```

keith-turner replied:
I had also thrown in that manager change, but that does not have to be done in this PR

brianloss replied:
Oops, yeah I glossed over the master/manager change too. Makes more sense the way you have it now.

brianloss commented on ansible/roles/accumulo/templates/accumulo-cluster-systemd:
```diff
@@ -287,26 +292,35 @@ function stop_all() {
 
   for host in "${hosts_to_check[@]}"; do
-    for svc in  tserver gc master monitor tracer; do
+    for svc in  tserver gc master monitor; do
       stop_service "$host" "$svc"
     done
+{% if accumulo_version is version('2.1.0','<') %}
+    stop_service "$host" "tracer"
+{% endif %}
```
This could be kept together too. It's probably just preference in the end over what looks better to you.
```suggestion
    for svc in  tserver gc master monitor{{ ' tracer' if accumulo_version is version('2.1.0','<') }}; do
      stop_service "$host" "$svc"
    done
```

brianloss reviewed (approved):
LGTM. I made a couple of suggestions for potential changes, but what's there looks correct to me and my suggestions are just to potentially improve readability of the source and/or generated file.

keith-turner commented:
The build is failing.  May be running into this issue https://github.com/ansible-community/ansible-lint/issues/1795

arvindshmicrosoft commented:
> The build is failing. May be running into this issue [ansible-community/ansible-lint#1795](https://github.com/ansible-community/ansible-lint/issues/1795)

Strange... the Python version is still 3.9.9, so that has not changed. Anyways looks like bumping the version of ansible-lint is a simple fix.

keith-turner commented:
> Strange... the Python version is still 3.9.9, so that has not changed. Anyways looks like bumping the version of ansible-lint is a simple fix.

Not completely sure, but I think the new version of ansible lint may have the following which fixes the issue with breaking change in one of its dependencies.

https://github.com/ansible-community/ansible-lint/pull/1797

arvindshmicrosoft commented:
> > Strange... the Python version is still 3.9.9, so that has not changed. Anyways looks like bumping the version of ansible-lint is a simple fix.
> 
> Not completely sure, but I think the new version of ansible lint may have the following which fixes the issue with breaking change in one of its dependencies.
> 
> [ansible-community/ansible-lint#1797](https://github.com/ansible-community/ansible-lint/pull/1797)

I'll double-check.

arvindshmicrosoft commented on ansible/accumulo.yml:
```diff
@@ -53,26 +53,34 @@
+    - name: "install and start all the accumulo tracer service using systemd"
+      block:
+        - import_tasks: roles/accumulo/tasks/start-tracer.yml
+      when: use_systemd and accumulo_version is version('2.1.0','<')
```
@keith-turner - sorry, I just spotted this, should it not be:
```suggestion
      when: accumulo_version is version('2.0.0','>=') and use_systemd and accumulo_version is version('2.1.0','<')
```
We really don't block someone from setting use_systemd to `True` when using Accumulo 1.x so I guess it is safer to filter this way.
to be consistent?

arvindshmicrosoft reviewed (approved):
@keith-turner - I just left a comment which would be a good-to-do item. Otherwise, it LTGM.

FYI, I tested Accumulo 2.0.1 (systemd and non-systemd) and 1.10.1 cluster deployments, including the `wipe` command, for good measure. All checked out okay.
